### PR TITLE
EditorConfig Support

### PR DESCRIPTION
**Pull request by tarik-brown**
Initial EditorConfig support for the TypeScript Side

- Parses .editorconfig file for document, onType and range formatting 
- Caches .editorconfig settings so that the file isnt parsed on each format operation

### CODE REVIEW

File: Extension/package.json [hunk 1]
```diff
@@ -157,10 +157,24 @@
...
         },
-        "C_Cpp.vcFormat.indent.preserveIndentationWithinParentheses": {
+        "C_Cpp.vcFormat.indent.withinParentheses": {
+          "type": "string",
+          "enum": [
+            "alignToPatenthesis",
+            "indent"
+          ],
+          "enumDescriptions": [
+            "%c_cpp.configuration.vcFormat.indent.withinParentheses.alignToPatenthesis.description%",
```
Comment (bobbrow) [inline Extension/package.json]: [sp] alignToPatenthesis

File: Extension/package.json [hunk 2]
```diff
@@ -157,10 +157,24 @@
           "description": "%c_cpp.configuration.vcFormat.indent.multiLineRelativeTo.description%",
           "scope": "resource"
         },
-        "C_Cpp.vcFormat.indent.preserveIndentationWithinParentheses": {
+        "C_Cpp.vcFormat.indent.withinParentheses": {
+          "type": "string",
+          "enum": [
+            "alignToPatenthesis",
```
Comment (bobbrow) [inline Extension/package.json]: [sp] alignToPatenthesis

File: Extension/src/LanguageServer/client.ts [hunk 3]
```diff
@@ -1086,9 +1086,18 @@ export class DefaultClient implements Client {
                         public provideOnTypeFormattingEdits(document: vscode.TextDocument, position: vscode.Position, ch: string, options: vscode.FormattingOptions, token: vscode.CancellationToken): Promise<vscode.TextEdit[]> {
                             return new Promise<vscode.TextEdit[]>((resolve, reject) => {
                                 this.client.notifyWhenReady(() => {
+                                    const settings: CppSettings = new CppSettings();
+                                    if (settings.formattingEngine !== "vcFormat") {
+                                        // If not using vcFormat, don't process on-type requests
+                                        // for '{' or '\n'
```
Comment (bobbrow) [inline Extension/src/LanguageServer/client.ts]: I was thinking we only did onType for clang-format with `}`.  I believe `;` should be filtered out too.

Comment (Colengms) [reply Extension/src/LanguageServer/client.ts]: I think you got this backwards.  Looking back at code prior to formatting changes, we only registered for `;`.  Also, the onType test automation currently only checks for `;`.

Comment (bobbrow) [reply Extension/src/LanguageServer/client.ts]: I take it back then. Should have looked at the code first. 😄

**Review by sean-mcmanus — APPROVED** [review]

**Review by sean-mcmanus — APPROVED** [review]

**Review by sean-mcmanus — CHANGES_REQUESTED** [review]
Is format disabling handled?

**Comment by Colengms** [comment]
> Is format disabling handled?

That's handled on the native side.  I could unregister the providers when disabled.  That would be a little more efficient.

**Review by sean-mcmanus — APPROVED** [review]

**Pull request merged**